File: ts/src/testutils/response_writer.ts
import { HttpHeader } from "../types/http_header";
import { HttpResponseWriter } from "../types/http_response_writer";

export class MockResponseWriter implements HttpResponseWriter {
  readonly _header: HttpHeader = new HttpHeader();
  StatusCode = 0;
  Body: string;

  constructor() {
    this.Body = "";
  }

  Header(): HttpHeader {
    return this._header;
  }
  WriteHeader(statusCode: number) {
    this.StatusCode = statusCode;
  }

  Write(data: string | Uint8Array): Promise<number> {
    const startLen = this.Body.length;
    if (typeof data !== "string") {
      this.Body = this.Body + new TextDecoder().decode(data);
    } else {
      this.Body = this.Body + data;
    }
    return Promise.resolve(this.Body.length - startLen);
  }
  End(): Promise<void> {
    return Promise.resolve();
  }
}

File: go/testutils/response_writer.go
package testutils

import (
	"bytes"
	"net/http"
)

type MockResponseWriter struct {
	StatusCode int
	HeaderRef  http.Header
	Body       *bytes.Buffer
}

func NewMockResponseWriter() *MockResponseWriter {
	return &MockResponseWriter{
		HeaderRef: http.Header{},
		Body:      bytes.NewBuffer([]byte{}),
	}
}

func (w *MockResponseWriter) Header() http.Header {
	return w.HeaderRef
}

func (w *MockResponseWriter) Write(data []byte) (int, error) {
	return w.Body.Write(data)
}

func (w *MockResponseWriter) WriteHeader(statusCode int) {
	w.StatusCode = statusCode
}
